presto-main/src/main/java/io/prestosql/operator/scalar/JoniRegexpFunctions.java

```diff
@@ -307,4 +308,101 @@ private static void validateGroup(long group, Region region)
+    @LiteralParameters("x")
+    @SqlType(StandardTypes.INTEGER)
+    public static long regexpPosition(@SqlType("varchar(x)") Slice source,
+                                      @SqlType(JoniRegexpType.NAME) JoniRegexp pattern,
+                                      @SqlType(StandardTypes.INTEGER) long start,
+                                      @SqlType(StandardTypes.INTEGER) long occurrence)
+    {
+        // start should bigger than 0
+        if (start < 1) {
+            throw new PrestoException(INVALID_FUNCTION_ARGUMENT, "start position cannot small than 1");
+        }
+        if (occurrence < 1) {
+            throw new PrestoException(INVALID_FUNCTION_ARGUMENT, "occurrence cannot small than 1");
+        }
+        // return -1 if start is bigger than the length of source
+        if (start > SliceUtf8.countCodePoints(source)) {
+            return -1;
+        }
+
+        Matcher matcher = pattern.matcher(source.getBytes());
+        long count = 0;
+        // convert char position to byte position
+        // subtract 1 because codePointCount starts from zero
+        int nextStart = SliceUtf8.offsetOfCodePoint(source, (int) start - 1);
+        while (true) {
+            int offset = matcher.search(nextStart, source.length(), Option.DEFAULT);
+            // Check whether offset is negative, offset is -1 if no pattern was found or -2 if process was interrupted
+            if (offset < 0 || offset >= source.length()) {
+                return -1;
+            }
+
+            if (++count == occurrence) {
+                // Plus 1 because position returned start from 1
+                return SliceUtf8.countCodePoints(source, 0, matcher.getBegin()) + 1;
+            }
+
+            if (matcher.getEnd() == matcher.getBegin()) {
+                nextStart = matcher.getEnd() + SliceUtf8.lengthOfCodePointFromStartByte(source.getByte(matcher.getBegin()));
+            }
+            else {
+                nextStart = matcher.getEnd();
+            }
+        }
+    }
+
+    @ScalarFunction
+    @Description("returns the number of times that a pattern occurs in a string")
+    @LiteralParameters("x")
+    @SqlType(StandardTypes.INTEGER)
+    public static long regexpCount(@SqlType("varchar(x)") Slice source, @SqlType(JoniRegexpType.NAME) JoniRegexp pattern)
+    {
+        Matcher matcher = pattern.matcher(source.getBytes());
+
+        int count = 0;
+        // Start from zero, implies the first byte
+        int nextStart = 0;
+        while (true) {
+            // Mather bug: search should return -1 if start equals length of source but start actually
+            int offset = matcher.search(nextStart, source.length(), Option.DEFAULT);
+            if (offset < 0 || offset >= source.length()) {
+                break;
+            }
+
+            if (matcher.getEnd() == matcher.getBegin()) {
+                nextStart = matcher.getEnd() + SliceUtf8.lengthOfCodePointFromStartByte(source.getByte(matcher.getBegin()));
```

Comment:
just use `nextStart = matcher.getEnd() + 1;`

Reply:
Can't use `matcher.getEnd() +1` here because I have to return code counts even if the pattern is empty.

Reply:
This is a bug in other JONI regex functions, see: https://github.com/prestosql/presto/pull/2313